File: backend/app/circuit_breaker.py
import time
from dataclasses import dataclass
from typing import Optional


@dataclass
class CircuitBreakerConfig:
    failure_threshold: int = 3
    success_threshold: int = 2
    timeout_seconds: int = 60


class CircuitBreaker:
    def __init__(self, config: Optional[CircuitBreakerConfig] = None):
        self.config = config or CircuitBreakerConfig()
        self.state = "closed"
        self.consecutive_failures = 0
        self.consecutive_successes = 0
        self.opened_at = 0.0

    def allow_request(self) -> bool:
        if self.state == "closed":
            return True
        if self.state == "half_open":
            return True
        if time.time() - self.opened_at >= self.config.timeout_seconds:
            self.state = "half_open"
            self.consecutive_successes = 0
            return True
        return False

    def record_success(self) -> None:
        self.consecutive_failures = 0
        if self.state == "half_open":
            self.consecutive_successes += 1
            if self.consecutive_successes >= self.config.success_threshold:
                self.state = "closed"
                self.consecutive_successes = 0
        else:
            self.state = "closed"

    def record_failure(self) -> None:
        self.consecutive_successes = 0
        self.consecutive_failures += 1
        if self.consecutive_failures >= self.config.failure_threshold:
            self.state = "open"
            self.opened_at = time.time()

    def snapshot(self) -> dict:
        return {
            "state": self.state,
            "consecutiveFailures": self.consecutive_failures,
            "consecutiveSuccesses": self.consecutive_successes,
            "openedAt": self.opened_at,
        }

File: worker/src/upstream/circuitBreaker.ts
// Circuit breaker — mirrors backend/app/circuit_breaker.CircuitBreaker.
// Per-provider (provider_id:model) breaker with closed/open/half_open states.

import type { CircuitSnapshot } from "../types/provider";

export interface CircuitBreakerConfig {
  failureThreshold: number;
  successThreshold: number;
  timeoutSeconds: number;
}

export const DEFAULT_BREAKER_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 3,
  successThreshold: 2,
  timeoutSeconds: 60,
};

type BreakerState = "closed" | "open" | "half_open";

export class CircuitBreaker {
  state: BreakerState = "closed";
  consecutiveFailures = 0;
  consecutiveSuccesses = 0;
  openedAt = 0;
  private readonly now: () => number;

  constructor(
    private readonly config: CircuitBreakerConfig = DEFAULT_BREAKER_CONFIG,
    now?: () => number,
  ) {
    this.now = now ?? (() => Date.now() / 1000);
  }

  allowRequest(): boolean {
    if (this.state === "closed") return true;
    if (this.state === "half_open") return true;
    if (this.now() - this.openedAt >= this.config.timeoutSeconds) {
      this.state = "half_open";
      this.consecutiveSuccesses = 0;
      return true;
    }
    return false;
  }

  recordSuccess(): void {
    this.consecutiveFailures = 0;
    if (this.state === "half_open") {
      this.consecutiveSuccesses += 1;
      if (this.consecutiveSuccesses >= this.config.successThreshold) {
        this.state = "closed";
        this.consecutiveSuccesses = 0;
      }
    } else {
      this.state = "closed";
    }
  }

  recordFailure(): void {
    this.consecutiveSuccesses = 0;
    this.consecutiveFailures += 1;
    if (this.consecutiveFailures >= this.config.failureThreshold) {
      this.state = "open";
      this.openedAt = this.now();
    }
  }

  snapshot(): CircuitSnapshot {
    return {
      state: this.state,
      consecutiveFailures: this.consecutiveFailures,
      consecutiveSuccesses: this.consecutiveSuccesses,
      openedAt: this.openedAt,
    };
  }
}
